File: src/main/java/com/binarysushi/studio/projectWizard/StudioProjectGenerator.java
package com.binarysushi.studio.projectWizard;

import com.binarysushi.studio.StudioBundle;
import com.binarysushi.studio.StudioIcons;
import com.binarysushi.studio.configuration.StudioConfigurationPanel;
import com.binarysushi.studio.configuration.StudioConfigurationProvider;
import com.intellij.ide.util.projectWizard.ModuleBuilder;
import com.intellij.ide.util.projectWizard.SettingsStep;
import com.intellij.ide.util.projectWizard.WebProjectTemplate;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.platform.ProjectGeneratorPeer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class StudioProjectGenerator extends WebProjectTemplate<StudioProjectWizardData> {

    @NotNull
    @Override
    public String getName() {
        return StudioBundle.message("studio.project.name");
    }

    @Override
    public String getDescription() {
        return StudioBundle.message("studio.project.description");
    }

    @Override
    public Icon getIcon() {
        return StudioIcons.STUDIO_ICON;
    }

    @Override
    public void generateProject(@NotNull Project project, @NotNull VirtualFile baseDir, @NotNull StudioProjectWizardData data, @NotNull Module module) {
        StudioConfigurationProvider configurationProvider = StudioConfigurationProvider.getInstance(project);
        configurationProvider.setHostname(data.hostname);
        configurationProvider.setUsername(data.username);
        configurationProvider.setPassword(data.password);
        configurationProvider.setVersion(data.version);
        configurationProvider.setAutoUploadEnabled(data.autoUploadEnabled);
    }

    @NotNull
    @Override
    public ProjectGeneratorPeer createPeer() {
        return new StudioProjectGeneratePeer();
    }

    @NotNull
    @Override
    public ModuleBuilder createModuleBuilder() {
        return new StudioModuleBuilder();
    }

    @SuppressWarnings("deprecation")
    private class StudioProjectGeneratePeer implements GeneratorPeer {

        private final StudioConfigurationPanel myConfigurationPanel = new StudioConfigurationPanel();

        @NotNull
        @Override
        public JComponent getComponent() {
            return myConfigurationPanel.createPanel();
        }

        @Override
        public void buildUI(@NotNull SettingsStep settingsStep) { }

        @NotNull
        @Override
        public Object getSettings() {
            return new StudioProjectWizardData(
                    myConfigurationPanel.getHostname(),
                    myConfigurationPanel.getUsername(),
                    myConfigurationPanel.getPassword(),
                    myConfigurationPanel.getVersion(),
                    myConfigurationPanel.getAutoUploadEnabled());
        }

        @Nullable
        @Override
        public ValidationInfo validate() {
            return null;
        }

        @Override
        public boolean isBackgroundJobRunning() {
            return false;
        }

        @Override
        public void addSettingsStateListener(@NotNull SettingsStateListener listener) {

        }
    }
}

File: src/main/kotlin/com/binarysushi/studio/projectWizard/StudioProjectGenerator.kt
package com.binarysushi.studio.projectWizard

import com.binarysushi.studio.StudioBundle.message
import com.binarysushi.studio.StudioIcons
import com.binarysushi.studio.configuration.projectSettings.StudioConfigurationPanel
import com.binarysushi.studio.configuration.projectSettings.StudioConfigurationProvider
import com.intellij.ide.util.projectWizard.ModuleBuilder
import com.intellij.ide.util.projectWizard.SettingsStep
import com.intellij.ide.util.projectWizard.WebProjectTemplate
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.platform.ProjectGeneratorPeer
import javax.swing.Icon
import javax.swing.JComponent

class StudioProjectGenerator : WebProjectTemplate<StudioProjectWizardData?>() {
    override fun getName(): String {
        return message("studio.project.name")
    }

    override fun getDescription(): String? {
        return message("studio.project.description")
    }

    override fun getIcon(): Icon {
        return StudioIcons.STUDIO_ICON
    }

    override fun generateProject(
        project: Project,
        baseDir: VirtualFile,
        data: StudioProjectWizardData,
        module: Module
    ) {
        val configurationProvider = StudioConfigurationProvider.getInstance(project)
        configurationProvider.hostname = data.hostname
        configurationProvider.username = data.username
        configurationProvider.password = data.password
        configurationProvider.version = data.version
        configurationProvider.autoUploadEnabled = data.autoUploadEnabled
    }

    override fun createPeer(): ProjectGeneratorPeer<StudioProjectWizardData?> {
        return StudioProjectGeneratePeer()
    }

    override fun createModuleBuilder(): ModuleBuilder {
        return StudioModuleBuilder()
    }

    private inner class StudioProjectGeneratePeer : ProjectGeneratorPeer<StudioProjectWizardData?> {
        private val myConfigurationPanel = StudioConfigurationPanel()
        override fun getComponent(): JComponent {
            return myConfigurationPanel.createPanel()
        }

        override fun buildUI(settingsStep: SettingsStep) {}
        override fun getSettings(): StudioProjectWizardData {
            return StudioProjectWizardData(
                myConfigurationPanel.hostname,
                myConfigurationPanel.username,
                myConfigurationPanel.password,
                myConfigurationPanel.version,
                myConfigurationPanel.autoUploadEnabled
            )
        }

        override fun validate(): ValidationInfo? {
            return null
        }

        override fun isBackgroundJobRunning(): Boolean {
            return false
        }
    }
}
